spec/models/user_spec.rb

```diff
@@ -0,0 +1,90 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  describe '#new_with_session' do
+    let(:empty_session) { {} }
+    let(:session) do
+      {
+        'devise.facebook_data':
+          OmniAuth::AuthHash.new(
+            provider: 'facebook',
+            uid: '1234567890',
+            info: { email: 'example@example.com', id: '1234567890', name: 'Johnnie Walker' },
+            extra: { raw_info: { email: 'example@example.com' } }
+          )
+      }
+    end
+
+    context 'when user does not have assigned email' do
+      context 'when session contains email' do
+        it 'assigns email from session to user' do
+          user = User.new_with_session({}, session)
+          expect(user.email).to eq('example@example.com')
+        end
+      end
+
+      context 'when session does not contain email' do
+        it 'does not assign email to user' do
+          user = User.new_with_session({}, empty_session)
+          expect(user.email).to eq('')
+        end
+      end
+    end
+
+    context 'when user has assigned email' do
+      context 'when session contains email' do
+        it 'does not assign email to user' do
+          user = User.new_with_session({ email: 'first_email@example.com' }, session)
```

Comment:
you can extract `subject` and only change `params` and `session` in contexts - so they are defined where they are used.